examples/GroupCharacters.jl

```diff
@@ -0,0 +1,558 @@
+mod(t, 2)
+```
+"""
+module GroupCharacters
+
+using Oscar
+
+# character values are elements from QabField
+#using QabModule
+#import QabModule: QabElem
+include("QabAndPChars.jl")
+
+# functionality for displaying character tables
+include("MatrixDisplay.jl")
+
+import Base: getindex, length, mod, one, print, show, zero
+
+import AbstractAlgebra: nrows, ncols
+
+import Nemo: degree
+
+# Load GAP's character table library (if it is installed)
+GAP.Packages.load("ctbllib")
+
+
+#############################################################################
+##
+##  conversion between Julia's QabElem and GAP's cyclotomics
+##
+function QabModule.QabElem(cyc::Union{GAP.GapObj,Int64})
+    GAP.Globals.IsCyc(cyc) || error("cyc must be a GAP cyclotomic")
+    denom = GAP.Globals.DenominatorCyc(cyc)
+    n = GAP.Globals.Conductor(cyc)
+    coeffs = GAP.Globals.ExtRepOfObj(cyc * denom)
+    cycpol = GAP.Globals.CyclotomicPol(n)
+    GAP.Globals.ReduceCoeffs(coeffs, cycpol)
+    coeffs = GAP.gap_to_julia(coeffs, recursive = true)
+    coeffs = coeffs[1:(length(cycpol)-1)]
+    denom = GAP.gap_to_julia(denom)
+    F, z = Nemo.CyclotomicField(n)
+    val = Nemo.elem_from_mat_row(F, Nemo.matrix(Nemo.ZZ, [coeffs]), 1, fmpz(denom))
+    return QabModule.QabElem(val,n)
+end
+
+#function QabModule.QabElem(cyc::GAP.GapObj, c::Int)
+#end
+
+function gap_cyclotomic(elm::QabModule.QabElem)
+    coeffs = [Nemo.coeff(elm.data, i) for i in 0:(elm.c-1)]  # fmpq
+    coeffs = [GAP.julia_to_gap(BigInt(numerator(x)) // BigInt(denominator(x)))
+              for x in coeffs]
+    return GAP.Globals.CycList(GAP.julia_to_gap(coeffs))
+end
+
+
+#############################################################################
+##
+##  character tables
+##
+abstract type GroupCharacterTable end
+
+mutable struct GAPGroupCharacterTable <: GroupCharacterTable
+    GAPGroup::Oscar.GAPGroup    # the underlying group, if any
+    GAPTable::GAP.GapObj  # the character table object
+    characteristic::Int
+    AbstractAlgebra.@declare_other
+
+    function GAPGroupCharacterTable(G::Oscar.GAPGroup, tab::GAP.GapObj, char::Int)
+      ct = new()
+      ct.GAPGroup = G
+      ct.GAPTable = tab
+      ct.characteristic = char
+      return ct
+    end
+
+    function GAPGroupCharacterTable(tab::GAP.GapObj, char::Int)
+      ct = new()
+      #ct.GAPGroup is left undefined
+      ct.GAPTable = tab
+      ct.characteristic = char
+      return ct
+    end
+end
+
+function character_table(G::Oscar.GAPGroup, p::Int = 0)
+    tbl = AbstractAlgebra.get_special(G, :character_table)
+    if tbl == nothing
+      tbl = GAP.Globals.CharacterTable(G.X)
+      AbstractAlgebra.set_special(G, :character_table => tbl)
+    end
+
+    if p != 0
+      isprime(p) || error("p must be 0 or a prime integer")
+      tbl = mod(tbl, p)
+    end
+
+    return GAPGroupCharacterTable(G, tbl, p)
+end
+
+function character_table(id::String, p::Int = 0)
+    if GAP.Globals.IsBoundGlobal(GAP.julia_to_gap("CTblLib"))
```

Comment:
I would recommend to concentrate input validation and availability checks at the start of the function, i.e., exit early if some constraint is not satisfied. That reduces the need for code indentation and IMHO also makes it easier to read and understand the code. So something like this, perhaps:
```suggestion
    GAP.Globals.IsBoundGlobal(GAP.julia_to_gap("CTblLib")) || error("no character table library available")
```
or
```suggestion
    hasproperty(GAP.Globals, :CTblLib) || error("no character table library available")
```